File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.*;
class Solution {
    static class Pair{
        char f; int s;
        Pair(char f, int s){
            this.f = f;
            this.s = s;
        }
    }
    static void calc(Pair[] pArr, char c, int score){
        for(Pair p : pArr){
            if(p.f == c) {
                p.s += score;
                break;
            }
        }
    } 
    public String solution(String[] survey, int[] choices) {
        String ret = "";
        Pair[] pArr = new Pair[8];
        pArr[0] = new Pair('R', 0);
        pArr[1] = new Pair('T', 0);
        pArr[2] = new Pair('C', 0);
        pArr[3] = new Pair('F', 0);
        pArr[4] = new Pair('J', 0);
        pArr[5] = new Pair('M', 0);
        pArr[6] = new Pair('A', 0);
        pArr[7] = new Pair('N', 0);
        
        for(int i = 0; i < survey.length; i++){
            int score = choices[i];
            String s = survey[i];
            if(score < 4) {
                char c = s.charAt(0);
                score = 4 - score;
                calc(pArr, c, score);
            } else if(score > 4){
                char c = s.charAt(1);
                score %= 4;
                calc(pArr, c, score);
            }
        }
        // for(Pair p : pArr){
        //     System.out.println(p.f + "::" + p.s);
        // }
        for(int i = 0; i < 8; i += 2){
            ret += (pArr[i].s >= pArr[i + 1].s) ? pArr[i].f : pArr[i + 1].f;
        }
        return ret;
    }
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.py
def solution(survey, choices):
    listing = ['RT', 'CF', 'JM', 'AN']
    dic = {x:[0,0] for x in listing}
    
    for sv, ch in zip(survey, choices):
        if ch ==  4:
            continue
        idx = ch // 4
        key = ''.join(sorted(list(sv)))
        idx = key.index(sv[idx])
        score = ch % 4 if ch > 4 else (4-ch)
        dic[key][idx] += score
        
    ret = ''
    for alp in listing:
        c = ''
        if dic[alp][0] < dic[alp][1]:
            c = alp[1]
        else:
            c = alp[0]
        ret += c
    return ret
    
        

"""
동의 비동의는 계속 바뀜 방향바뀜
지표별 점수 높은 놈 => 사전순
각 알파벳 마다 점수 
점수는 4를 기준으로 큰놈은 오른쪽 작은 놈은 왼쪽에 점순 넣기
'AN': [0, 0]
배열에 순서대로 넣고 비교하면서 순회하면 되겠다
1. 조사를 돌면서 점수를 계산
2. 계산된 점수에 따라 정렬 
3. 성향 출력
"""